infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -111,43 +122,169 @@ private long getMaxReplicatedStreamsTail() {
+        long noRemainingEntriesToSend = 0;
+
+        // If we have processed up to or beyond the latest known tx stream tail
+        // we can assume we are up to date.
+        // Note: in the case of a switchover the tail won't be moving so we can assume
+        // the tx's stream last processed timestamp will never be above the stream's tail,
+        // but in the case of ongoing replication or holes, it might be above.
+        // We can't do much other than report that we're up to date (as it might continue moving)
+        if (txStreamTail <= currentTxStreamProcessedTs.getTimestamp()) {
+            // (Case 2 from description)
+            // If the current processed tx stream entry has data to replicate,
+            // we need to ensure we have received an ACK it, otherwise,
+            // we might be signalling completion when there is still an entry for which
+            // we haven't received confirmation of the recipient.
+            if (currentTxStreamProcessedTs.isStreamsToReplicatePresent()) {
+                if (lastAckedTs == currentTxStreamProcessedTs.getTimestamp()) {
+                    log.trace("");
+                    // (Case 2.0)
+                    return noRemainingEntriesToSend;
+                }
+
+                // (Case 2.2)
+                // Last ack'ed timestamp should match the last processed tx stream timestamp.
+                // Calculate how many entries are missing, based on the tx stream's address map
+                return getTxStreamTotalEntries(lastAckedTs, currentTxStreamProcessedTs.getTimestamp());
+            }
+
+            // (Case 2.1)
+            // Since last tx stream processed timestamp is not intended to be replicated
+            // and we're at or beyond the last known tail, no entries remaining to be sent
+            // at this point.
+            return noRemainingEntriesToSend;
         }
-        return remainingEntriesToSend;
+
+        // (Case 1 from description)
+        if (currentTxStreamProcessedTs.isStreamsToReplicatePresent()) {
+            // (Case 1.0)
+            return getTxStreamTotalEntries(lastAckedTs, txStreamTail);
+        }
+
+        // Case (1.1)
+        return getTxStreamTotalEntries(currentTxStreamProcessedTs.getTimestamp(), txStreamTail);
+    }
+
+    private long getTxStreamTotalEntries(long lowerBoundary, long upperBoundary) {
+        long totalEntries = 0;
+
+        if (upperBoundary > lowerBoundary) {
+            StreamAddressRange range = new StreamAddressRange(TRANSACTION_STREAM_ID, upperBoundary, lowerBoundary);
+            StreamAddressSpace txStreamAddressSpace = runtime.getSequencerView().getStreamAddressSpace(range);
+            // Count how many entries are present in the Tx Stream (this can include holes,
+            // valid entries and invalid entries), but we count them all (equal weight).
+            // An invalid entry, is a transactional entry with no streams to replicate (which will be ignored)
```

Comment:
where are we ignoring the invalid entry?

Reply:
In StreamsLogEntryReader isValidTransactionEntry method

Reply:
that entry will not get replicated but will get counted here, right?  We may want to update the comment if so?